textdb/textdb-dataflow/src/main/java/edu/uci/ics/textdb/dataflow/queryrewriter/FuzzyTokenizer.java

```diff
@@ -43,36 +44,106 @@ public FuzzyTokenizer(String phrase) {
+     */
+    public List<String> getFuzzyTokens() {
+        String[] terms = phrase.split("\\s+");
+        String rewrittenPhrase = "";
+        for(String term : terms)
+            rewrittenPhrase += " " + rewriteTerm(term);
+        rewrittenPhrase = rewrittenPhrase.substring(1);
+        return Arrays.asList(rewrittenPhrase);
+    }
+
+    /**
+     * Performs most likely tokenization of a single term
+     * Uses Dynamic Programming algorithm approach used for tokenization of Chinese strings
+     * Source - Chinese Tokenizer in SRCH2 codebase
+     *           https://github.com/SRCH2/srch2-ngn/blob/master/src/core/analyzer/ChineseTokenizer.cpp#L197
+     * @param term
+     * @return
+     */
+    public String rewriteTerm(String term) {
+        int size = term.length();
+        if(size == 0)
+            return "";
+
+        double[] scoreAtGap = new double[size];
+        int[] preBestGap = new int[size];
+
+        Arrays.fill(scoreAtGap, RewriterConstants.MAXIMUM_SEQUENCE_SCORE);
+        Arrays.fill(preBestGap, -1);
+        scoreAtGap[0] = 0;
+        preBestGap[0] = 0;
+
+        for(int endPosition = 1; endPosition < size; endPosition++) {
+            for(int spanSize = 1; spanSize < RewriterConstants.MAX_WORD_LENGTH && spanSize <= endPosition; spanSize++) {
+                double freq = wordBase.getFrequency(term.substring(endPosition-spanSize, endPosition+1));
+                if(freq < 0) {
+                    if(spanSize == 1)
+                        freq = RewriterConstants.UNKNOWN_CHAR_FREQ;
+                    else
+                        continue;
+                }
+
+                if(freq + scoreAtGap[endPosition-spanSize] + RewriterConstants.TOKEN_LENGTH_PENALTY < scoreAtGap[endPosition]) {
+                    scoreAtGap[endPosition] = freq + scoreAtGap[endPosition-spanSize] + RewriterConstants.TOKEN_LENGTH_PENALTY;
+                    preBestGap[endPosition] = endPosition - spanSize;
+                }
+            }
+        }
+
+        String tokenization = "";
+
+        for(int pos = size-1; pos > 0; pos = preBestGap[pos]-1) {
+            tokenization = " " + term.substring(preBestGap[pos], pos+1) + tokenization;
+        }
+
+        return tokenization.substring(1);
+    }
+
+    /**
      * When called returns array of rewritten phrases
      * Splits phrase string into delimited (by default, it is delimited using a space) terms
      * For each term performs fuzzy tokenization to create corresponding lists of rewritten terms
-     * Calls method rewriteTerm to perform tokenization over single term
+     * Calls method bruteRewriteTerm to perform tokenization over single term
      * Performs a cross product concatenation of each rewritten term list and returns final String array
      * Calls method crossCatenate to perform required concatenation of multiple lists
      * @return
      */
-    public List<String> getFuzzyTokens() {
+    public List<String> bruteGetFuzzyTokens() {
```

Comment:
The DP method only returns the most likely tokenization. The brute force method is the only method till now which returns all possible tokenizations and that's why we were thinking of still keeping it around till we find a substitute for that use-case.